Stac support

ashoka1234 opened the pull request:
## `STAC-API` for `datacube-explorer`
This is a minimalist implementation of `STAC-API` spec for `datacube-explorer`. It consists of a single endpoint `/stac/search`  where `HTTP POST` is a required implementation while `HTTP GET` is an equivalent optional implementation. The search fields `bbox` (bounding box) and `time` are required fields for `STAC-API` while `product` field is made required for our purposes. The endpoint returns a list of `STAC items` that matches the query parameters in a `GeoJson FeatureCollection` format. If the query has more datasets that could be sent in a single return, `paging` is supported via a `link` in the `links` field of `GeoJson FeatureCollection`. The `STAC-API` spec could be found [here](https://github.com/radiantearth/stac-spec/tree/master/api-spec).

There are no changes done to `datacube-explorer` front-end rendering aspects and therefore testing could be done directly from the browser and/or using `curl`. The `API` is implemented as a `Flask blueprint` in `_stac.py`.

omad committed:
Basic Stac support

jeremyh commented:
Thanks Ashoka.

Can you add some tests before we review it? Coveralls was broken because of the recent repository rename, so it hasn't complained here. -- I've just fixed it.

There's code that provides test data and configuration already: see the (basic) examples for testing the current geojson api and front-end pages: https://github.com/opendatacube/datacube-explorer/blob/c99aede051489d353595b8c7b53983076b0c49a9/integration_tests/test_page_loads.py#L172

ashoka1234 commented:
Thanks Jeremy.

Sure will add the tests.

ashoka1234 commented:
Added  single test that cover both `GET` and `POST`. If you needed further tests or more breath please let me know

jeremyh reviewed (dismissed):
Thanks Ashoka.

From running it locally, there’s a few errors coming back. It might be worth adding tests for them:

(several of these shouldn't throw server exceptions as they're user errors, and we’ll receive system notifications.)

- The “next” page link should be valid (test that the url can be loaded).
- User forgetting required fields. Ideally return an understandable error message to the user too, eg: https://github.com/opendatacube/datacube-explorer/blob/a57f72390868d298147b60d0282bcd2a389817a5/cubedash/_product.py#L18-L19
- A minimal query should work (eg. no “limit” specified by user)
- A query that returns no results (eg. out of the time/space bounds).
- A query for a product that uses geo_ref_points (I assume it's intended to be supported as there's some code for it).

(I'm excited about tests here, as they have a pretty high bang-for-the-buck: it’s a read-only api, publicly facing, with a strong specification, and there’s already test data and setup for it.)

I'm guessing the original prototype code was only written for a few specific products in the cloud instance, as it's returning blank values for most of the existing dea collection products. I know it's a minimal initial implementation, so its probably intentional, but we should make it clear if so. (even if that means returning an "unsupported product" error to the user, it's better than quiet invalid values).

Eg. The “assets” section paths are all blank in most products. ‘ls8_nbar_albers’:

![screen shot 2019-02-01 at 10 56 55 am](https://user-images.githubusercontent.com/25688/52093754-81c4c780-2610-11e9-8f83-163ee16c2bd3.png)

Also:
- Is the root page (`"/"`) usually blank in stac?
- Does stac have a jsonschema or something we could throw into a test ("return document validates against schema")? That could catch many issues.
- You mentioned 'product' is our own extension, but is required, so presumably we're not stac compliant just yet. The underlying datacube `index.search` methods should work without a product specified (they will union all products that match the search terms).

ashoka1234 commented:
Sure will address those. My bad! I missed parsing bbox!

ashoka1234 commented:
Hi Jeremy,
I have done changes and refactoring and followings are comments about those:

1. Projection code (to `epsg:4326`)  refactored so that it works with broad number of projection strings.
2. next page link code refactored so that it works with more edge cases.
3. `product` is no longer compulsory and associated refactoring done.
4. `bbox` parsing done which I missed earlier. 
5. Added user error messages instead of relying on system error messages.
6. Added tests for the two products that are populated for the test run (`wofs_albers` and `high_tide_comp_20p`) including testing `next` link

However, there are few issues that needed unfulfilled at this stage:
1. blank asset links: For datasets in `.nc` files, `metadata_doc` returns empty band path names. We don't have a clear solution to this at this stage.
2. There is a `STAC` validator that could be used to validate the entire `STAC item` but we have to avoid using it until we have a clear direction of how we overlay a conceptual `STAC catalog` on `datacube database` .
3. The root page is meant to be the optional `danamic STAC catalog` which is a substantial work item that overlap with above issues so hope to implement it in a separate PR.
4. `STAC API` spec is defined in terms of `openAPI` so I believe there could be schema validators out there but not sure.  

Please let me know if you see further changes that you see good to have.

jeremyh committed:
WIP catalog prototype

jeremyh committed:
Further fixes, additions to catalog prototype

jeremyh reviewed (dismissed):
Thanks Ashoka, the next link loads and it’s good to see a load test. Please add tests when you fix other issues too! For instance: remove “limit” from one of the test requests, so that we know it now works without a limit. There were a few other cases I mentioned in the previous review that weren't/aren't working and aren't tested.

Note that the testing can be simpler/easier than it is now — see [the existing test I pointed to earlier](https://github.com/opendatacube/datacube-explorer/blob/d3ac35b750a2fdf4402e90554f5de29c0c18df5b/integration_tests/test_page_loads.py#L184-L208). It does almost the same thing that we want to do for stac: 
- it's hitting urls for a geojson response. 
- it's checking that the return was successful, via a helper method, which the current stac test doesn't.
- the test itself is querying the same test data, in a similar way, so the returned data counts will be the same as our stac api.

Feel free to copy it! (and improve it — stac has paging, spatial bounds, etc)

There are other tests in the same file that might be worth adapting too, eg:
https://github.com/opendatacube/datacube-explorer/blob/d3ac35b750a2fdf4402e90554f5de29c0c18df5b/integration_tests/test_page_loads.py#L209-L224

I'm still hitting a lot of bugs in the basic behaviour (simple searches+limits+paging), and it dies with exceptions on many of our products. Add some tests that use the limits and paging. For instance: does the correct total count of datasets come back across all pages? What if the limit/page is smaller/larger than results? And when some query params are omitted? There's only a small amount of test data, so we can be exact in our asserts. It's quite error-prone logic, and we've already seen many bugs.

The exceptions thrown seem to be related to the lat/lon extraction code. Rather than fix this code, it might be better to change from using `datasets.search_returning()` to `datasets.search()`. This gives us Dataset objects, which have bounds already, so we won't have to calculate them ourselves and handle the edge cases.

After reading the stac docs I think several of the fields are incorrect too. We can discuss that separately once the basic querying is working -- you mentioned already that there's planned work to map our internal structures onto stac.

ashoka1234 commented:
Hi Jeremy,
I have added hard a limit and updated tests for testing GET, POST, links, limit, product, and assets. Also note that I have added JSON schema validator of STAC assets to tests as well. The code that handle the limit is refactored so that it works when no product is specified.

ashoka1234 commented:
Hi Jeremy, I may have now added all the tests you wanted

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+
+@bp.route('/')
+def root():
+    return utils.as_json({})
```
A blank response might be confusing to users -- it looks like a mistake, and it may confuse some stac clients.

A 404 error could more explicitly tell them that a root catalog is not support yet:
```suggestion
    abort(404, "Only /stac/search is currently supported")
```
(perhaps a 501-not-implemented error is more correct, but it's cacheable as a permanent response, which doesn't sound wise to me. All other unimplemented paths are 404s by default, so I'm leaning towards it.)

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+
+    # bbox and time are compulsory
+    if not bbox:
+        abort(404, "bbox must be specified")
```
Use status code 400 for bad arguments, not 404. (there's a few more in this file to change too)
```suggestion
        abort(400, "bbox must be specified")
```

See the list of different codes here: https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
(or more friendly: https://httpstatuses.com/)

(Note that `bbox` and `time` are optional in the stac spec, but we can implement that later.)

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+    # from_dts must be lower than limit
+    limit = MAX_DATASETS if not limit else min(limit, MAX_DATASETS)
+
+    if from_dts and from_dts >= limit:
```
?
```suggestion
    if from_dts and from_dts >= MAX_DATASETS:
```

ashoka1234 replied:
I wanted to keep `from_dts` within `limit` specified so that `paging` occur only within `limit` . The reason was that we anyway truncate at `MAX_DATASETS` . This could be seen as deviating from the spec but `MAX_DATASETS` also forces us to deviate. Not really sure about this!

ashoka1234 replied:
Wanted to keep paging within `limit` specified since anyway we are truncating at `MAX_DATASETS`

jeremyh replied:
I think this is breaking from the stac spec in a way that will confuse users, and I can't see any benefits to doing it this way (it's not any simpler, right?).

If I understand it correctly:
- The stac spec behaviour is that specifying `limit=3`, I'll get 3 results on a page, and I can keep loading pages until we hit our own internal `MAX_DATASETS` limit (which we may remove in the future).
- With this behaviour, if I query with `limit=3`, I'll get 3 results and a later page will throw an error because it's higher than 3. Stac clients that already specify their own page sizes may not realise they're limiting the total datasets they can see. Worse, they will need to change it to 10000 to see all 10000 datasets up to the `MAX_DATASETS` limit (which we don't want them to start doing!).

If I really want to limit myself to 3 results total, I can already do that by setting `limit=3` and  ignoring the next page; I don't need the second behaviour to do it for me.

ashoka1234 replied:
Hmm OK I will change this. I had interpreted `limit` as total required rather than a possible page size

ashoka1234 replied:
done

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+    result["features"] = [stac_dataset(dt) for dt in datasets_]
+
+    if len(list(itertools.islice(datasets, 1))) == 1 and to_dts < limit:
+        url_next = host_url + 'stac/search?'
```
I doesn't really matter for this pull request, but we could use Flask's `url_for()` in the future to simplify this url code.

ashoka1234 replied:
Done

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+        query['time'] = Range(Timestamp(time_period[0]), Timestamp(time_period[1]))
+    elif len(time_period) == 1:
+        query['time'] = Range(Timestamp(time_period[0]).floor('1D'),
+                              (Timestamp(time_period[0]) + DateOffset(1)).floor('1D'))
```
The `floor()`ing here doesn't look right to me. If I search using `time=2018-02-12T23:20:50Z` it will return datasets of completely different times on the same day.

I do like that a value without time ("2017-04-31") will conveniently search a whole day range though.

The stac spec doesn't clarify this case as far as I can see. Taken literally it it currently requires time:

> 
> - A date-time: "2018-02-12T23:20:50Z"
> - A period: "2018-02-12T00:00:00Z/2018-03-18T12:31:12Z" or "2018-02-12T00:00:00Z/P1M6DT12H31M12S"
> 
> Only features that have a temporal property that intersects the value of time are selected.

The example implementation from the stac repo seems to support day-only as a range: https://sat-api.developmentseed.org/stac/search?datetime=2019-02-18

But ignores seconds (& milliseconds) when matching exact times: https://sat-api.developmentseed.org/stac/search?datetime=2019-02-18T00:37:51.660000+00:00
(I'm not sure whether we should copy this behaviour)

(I'm also slightly averse to using `pandas` here because it's _very_ lenient with accepted date formats, and we'll need to continue supporting all of those other date variations in the future if people start using them. Pandas also isn't currently declared as a dependency)

ashoka1234 replied:
done

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+    stac_item = OrderedDict([
+        ('id', metadata_doc['id']),
+        ('type', 'Feature'),
+        ('bbox', [metadata_doc['extent']['coord']['ll']['lon'],
```
Unfortunately we can't assume which of `ll`/`ur`/`ul`/`lr` have the highest/lowest values -- it will break on many existing datasets in our collections.

As I mentioned in the previous review, it's less error prone to use the Dataset's existing extent code (`dataset.extent`) for all of this, as it handles more edge cases already.

There's also a local utils method that handles a few other edge cases:
```
    from cubedash import _utils
    shape, is_valid_extent = _utils.dataset_shape(dataset)
    bbox = shape.bounds
```

(Note that `shape` may be null if the extent is missing or invalid and not repairable)

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+            band_name: {
+                # "type"? "GeoTIFF" or image/vnd.stac.geotiff; cloud-optimized=true
+                'href': band_data['path'],
+                "required": 'true',
```
is `required: true` a stac thing? I can't see it anywhere. What does it mean?

ashoka1234 replied:
I have carried this thing over from earlier work. Either it was present in earlier STAC versions or Siva may have added it for some other reason. I have lost track of this may be I will remove it.

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+                # "type"? "GeoTIFF" or image/vnd.stac.geotiff; cloud-optimized=true
+                'href': band_data['path'],
+                "required": 'true',
+                "type": "GeoTIFF"
```
`type: GeoTIFF` will be wrong on most of our products (and the [stac spec](https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md#media-types) says to use `image/tiff or image/vnd.stac.geotiff`*).

A missing field is better than a wrong field. We should either sniff the file extension or remove it for now, I think.

\* the example `sat-api.developmentseed.org` seems to use a third one: `image/x.geotiff`

ashoka1234 replied:
Not hard to sniff but I will remove it

jeremyh commented on integration_tests/test_page_loads.py:
```diff
@@ -383,3 +384,87 @@ def test_with_timings(client: FlaskClient):
+        assert len(geojson.get('features')) == DATASETS_PER_REQUEST
+        dataset_count += DATASETS_PER_REQUEST
+        next_links = [link['href'] for link in geojson.get('links', []) if link['rel'] == 'next']
+    assert dataset_count == MAX_DATASETS
```
Nice :+1:, testing that all pages add up!

jeremyh committed:
playing around with catalogue

jeremyh commented on integration_tests/test_page_loads.py:
```diff
@@ -383,3 +384,87 @@ def test_with_timings(client: FlaskClient):
+                                                'limit': DATASETS_PER_REQUEST}),
+                               headers={'Content-Type': 'application/json', 'Accept': 'application/json'})
+    geodata = json.loads(rv.data)
+    assert len(geojson.get('features')) <= DATASETS_PER_REQUEST
```
Make this an exact equality. We should know how much of our test data matches the query. (Ten, I think. But the limit is 5. Perhaps raise the limit to make the test more useful?)

There's also a subtle typo: `geojson` was the variable for the previous test's response, not this one (it should be `geodata`). We could split these paragraphs into separate test methods to make this mistake less likely.

ashoka1234 replied:
That was sloppy. Done

jeremyh commented on integration_tests/test_page_loads.py:
```diff
@@ -383,3 +384,87 @@ def test_with_timings(client: FlaskClient):
+        assert band in geodata['features'][0]['assets']
+
+    # Validate stac items with jsonschema
+    with open('integration_tests/schemas/stac_item.json', 'r') as fp:
```
This path will fail when the tests are run from a different directory.

See the geojson schema for an example that loads relative to the current file:

https://github.com/opendatacube/datacube-explorer/blob/1783ccf830d95357e4e6ad8dc5657c1e70010663/integration_tests/asserts.py#L18-L19

ashoka1234 replied:
Done

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+_LOG = logging.getLogger(__name__)
+bp = Blueprint('stac', __name__, url_prefix='/stac')
+
+CFG = yaml.load(
```
It doesn't have to be in this pull request, but we should move these to the settings module before release. Other users running the code shouldn't have "Geoscience Australia", and our phone number, returned as their default organisation.

ashoka1234 replied:
Done

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+
+    result = dict()
+    result["type"] = "FeatureCollection"
+    res_bbox = _compute_bbox([dt.metadata_doc_without_lineage() for dt in datasets_])
```
We can use `dt.metadata_doc` as search results have no lineage already.

`dt.metadata_doc_without_lineage()` is a brand new method that'll give us a hard requirement on an unreleased version of datacube [and whose name/implementation I've just seen and am slightly hesitant about].

ashoka1234 replied:
done

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+
+    result["features"] = [stac_dataset(dt) for dt in datasets_]
+
+    if len(list(itertools.islice(datasets, 1))) == 1 and to_dts < limit:
```
`to_dts < MAX_DATASETS`?

(aside: the purpose of the first clause could be made more clear: such as by adding an `iterator_has_next()` util method or a comment).

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+
+def stac_dataset(dataset):
+    """
+    Returns a dict corresponding to a stac item catalog
```
```suggestion
    Returns a dict corresponding to a stac item
```

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+    Returns a dict corresponding to a stac item catalog
+    """
+    metadata_doc = dataset.metadata_doc_without_lineage()
+    uri = dataset.local_uri
```
`dataset.local_uri` will break the current "datasets on s3" use case, as it only returns `file://` urls.

For now it may be better to grab the first uri from the `dataset.uris` array, which includes both local and remote uris (or if enthusiastic, filter for remote uris first.). 

How we handle/translate uris correctly will be a bigger future task.

ashoka1234 replied:
Agreed

jeremyh commented on cubedash/_stac.py:
```diff
@@ -0,0 +1,235 @@
+        geodata = valid_coord_to_geojson({'type': 'Polygon', 'coordinates': points}, dataset.crs)
+
+    # Convert the date to add time zone.
+    center_dt = parse(metadata_doc['extent']['center_dt'])
```
`dataset.center_time` will work with more of our products, and is already a datetime.

ashoka1234 replied:
done

jeremyh reviewed (dismissed):
Thanks Ashoka -- the tests are much better now! :+1:

There's only a few small things remaining.

(I've left out most comments on the stac `Item` document properties as I think we'll fix them in a future pull request.)

jeremyh committed:
wip: merging old-api and stac code

jeremyh committed:
Merge branch 'develop' into stac-support

jeremyh committed:
Allow multipolygon after buffering.

extent.to_crs() can chop at the meridian.

jeremyh reviewed (changes requested):
Thanks for getting Travis sorted Ashoka.

I haven't done a line-by-line yet, but from a quick test it looks like there's a few regressions:
- Many requests are timing out for me. It looks like there's no longer a limit passed to the db query.
- the "next" link is often missing. (I assume this is because of the new dataset skipping, as the return count wont be "limit+1")

ashoka1234 commented:
Hi Jeremy,
Yes `limit` had to be taken out from the query since we no longer enforce the product and the limit value is not correct. Also since we filter invalid datasets, the limit would not be correct either. If the removal of `limit` effect response time we could put some larger value greater than `MAX_DATASETS`. Counting is done on validated datasets. Next link will not be present if we reach `MAX_DATASETS` or if validated dataset generator is empty. It could be timeout affecting the generator or I don't seem to be able to reproduce that context.

jeremyh committed:
Add wrs84 index of bbox search

jeremyh committed:
Move api json construction closer to view

jeremyh committed:
bbox index should be gist

jeremyh committed:
Move argument parsing/validation closer to view layer

jeremyh committed:
Allow larger page sizes up to the dataset limit

The previous behaviour only allowed 'limit' to be smaller
than the default page size, which isn't very useful.

jeremyh committed:
Simplify url construction code

jeremyh committed:
Remove incorrect bbox for item collections

bbox is for entire search, not current page of search. But we
can't calculate that efficiently.

The example stac sat-api just puts the user requested bbox in it,
which doesn't seem useful enough.

In the future we may be able to fill it in from cubedash's
cached statistics.

jeremyh committed:
Test readability: urls, only print on failures

jeremyh commented:
I'm doing some work to merge this stac logic with the existing geojson api, as long-term we wont need both, and so we may as well consolidate effort earlier.

ashoka1234 commented:
Yes I also think that's a good idea.

jeremyh committed:
Split large stac test into separate tests, module

jeremyh committed:
Update to latest stac schemas

jeremyh committed:
Merge stac and geojson api implementations

jeremyh committed:
Add stac collection, item pages, fix links

jeremyh committed:
Loosen bbox, time, limit

Using the new backend it's no longer in-memory.

And we can now fully support the old api functionality.

jeremyh committed:
Fix all-time-range test for old api

jeremyh committed:
Revert older api to lightweight search

Until stac implementation is more mature.

jeremyh committed:
Item search api polish

Require users to specify their search keys.

jeremyh committed:
Update tests: search is center time, not full range

The rest of the UI (from summaries) are all generated based on center time, so this is more consistent.

jeremyh committed:
paging tests: ensure ordered, everything exactly once

jeremyh committed:
Add script to ease schema update

jeremyh committed:
Minimum jsonschema is required for references

jeremyh committed:
Missing dataset should return 404

jeremyh committed:
Improve assertion failure messages (wrap context)

jeremyh committed:
Only include perf time when explicitly asked for

jeremyh committed:
Fix fetch-all test: time is optional

jeremyh committed:
Ensure valid shapes from stac

jeremyh committed:
Fix assert context for empty assert messages

jeremyh committed:
Add assertion for buffer not collapsing polygon

jeremyh committed:
Clean, separate DB shape handling

jeremyh committed:
Add failing tests for valid href values

jeremyh committed:
Expand Item properties on datasets

jeremyh committed:
Bump attrs version, needed by jsonschema

jeremyh committed:
Bump jsonschema version

jeremyh committed:
Discourage use of the "offset" param

We may do paging in a better way in the future. Renaming
it makes it clearer it's supposed to be an internal param,
not for users.

jeremyh committed:
Don't duplciate paths in assets. Start eo:bands

eo:bands is supposed to be numerical in stac: todo.

jeremyh committed:
Add an 'alternative' html link for stac datasets

A place for users to view the dataset (following the example
in the stac spec)

jeremyh committed:
Fix expected count for all-test-data

jeremyh committed:
Avoid logging gigantic `repr(summary)`

jeremyh committed:
Update tests for product eo metadata

jeremyh committed:
Lack of region code shouldn't be a warning

It's normal for many products

jeremyh committed:
Add test for loading "WFS" collection items

jeremyh committed:
count-all-pages test should survive infinite loops

jeremyh committed:
Disable exact-computed-bbox checks that differ between pythons

jeremyh committed:
Fix infinite paging in collection-items test

jeremyh committed:
Make stac endpoint info configurable

jeremyh committed:
Make stac page size limits configurable

jeremyh committed:
Make all stac URLS default to absolute URLS

jeremyh committed:
Return all dates in UTC

jeremyh committed:
Clean up stac content types, docstrings

jeremyh committed:
Minor query cleanup, make ordering optional

jeremyh committed:
Make schema updates more visible

jeremyh commented:
I've pushed the (big! sorry) merge of the two backends. It fixes many of the paging issues and adds more tests.

It also includes:
- simple implementations of the root `/stac`, product catalogs, and Item pages.
- many additions to the Item structure/properties.
     -  but there's still  a lot more of our metadata we could map across. We probably need to discuss this and test it on more real-world datasets.
- all search fields (`bbox`/`time`) are now optional. They needed to be for the other api.
- the search now uses Explorer's spatial table.

Two areas that still need improvement:
- error messages should be returned as json (as defined in stac). We return http error codes, but only plain-text errors.
- `eo:bands` implementation on Items/Collections. I started this, but my `eo:bands` arrays currently contain strings (which datacube uses for band identification), and stac expects them to be mapped to integers.

... but I think it's good enough to merge this pull request as an MVP. We can make further improvements in (hopefully smaller sized) future PRs.

jeremyh committed:
Merge branch 'develop' into stac-support
